File: handlers/hostinfo/iops.go
package hostInfo

import (
	"encoding/json"
	"github.com/Sirupsen/logrus"
	"io/ioutil"
	"os"
	"runtime"
	"strconv"
)

type IopsCollector struct {
	data map[string]interface{}
}

func (i IopsCollector) getIopsData(readOrWrite string) (map[string]interface{}, error) {
	file, err := os.Open("/var/lib/rancher/state/" + readOrWrite + ".json")
	if err != nil {
		logrus.Error(err)
		return map[string]interface{}{}, err
	}
	data, _ := ioutil.ReadAll(file)
	var result map[string]interface{}
	json.Unmarshal(data, result)
	return result, nil
}

func (i IopsCollector) parseIopsData() map[string]interface{} {
	data := map[string]interface{}{}
	readJSONData, err1 := i.getIopsData("read")
	writeJSONData, err2 := i.getIopsData("write")
	if err1 != nil || err2 != nil {
		return data
	}
	readIops, _ := getFieldsIfExist(readJSONData["jobs"].([]map[string]interface{})[0], "read", "iops")
	writeIops, _ := getFieldsIfExist(writeJSONData["jobs"].([]map[string]interface{})[0], "write", "iops")
	device, _ := getFieldsIfExist(readJSONData["disk_util"].([]map[string]interface{})[0], "name")
	key := "/dev" + strconv.QuoteToASCII(device.(string))
	data[key] = map[string]interface{}{
		"read":  readIops,
		"write": writeIops,
	}
	return data
}

func (i IopsCollector) GetData() map[string]interface{} {
	if runtime.GOOS == "linux" {
		if len(i.data) == 0 {
			i.data = i.parseIopsData()
		}
		return i.data
	}
	return map[string]interface{}{}
}

func (i IopsCollector) getDefaultDisk() string {
	data := i.GetData()
	if len(data) == 0 {
		return ""
	}
	for key := range data {
		return key
	}
	return ""
}

func (i IopsCollector) KeyName() string {
	return "iopsInfo"
}

func (i IopsCollector) GetLabels(prefix string) map[string]string {
	return map[string]string{}
}

File: tests/tests/cattle/plugins/host_info/iops.py
import logging
import platform
import json

log = logging.getLogger('iops')


class IopsCollector(object):
    def __init__(self):
        self.data = {}

    def _get_iops_data(self, read_or_write):
        with open('/var/lib/rancher/state/' + read_or_write + '.json') as f:
            return json.load(f)

    def _parse_iops_file(self):
        data = {}

        try:
            read_json_data = self._get_iops_data('read')
            write_json_data = self._get_iops_data('write')
        except IOError:
            # File doesn't exist. Silently skip.
            return {}

        read_iops = read_json_data['jobs'][0]['read']['iops']
        write_iops = write_json_data['jobs'][0]['write']['iops']
        device = read_json_data['disk_util'][0]['name']
        key = '/dev/' + device.encode('ascii', 'ignore')
        data[key] = {'read': read_iops, 'write': write_iops}
        return data

    def key_name(self):
        return "iopsInfo"

    def get_data(self):
        if platform.system() == 'Linux':
            if not self.data:
                self.data = self._parse_iops_file()
            return self.data
        else:
            return {}

    def get_default_disk(self):
        data = self.get_data()
        if not data:
            return None

        # Return the first item in the dict
        return data.keys()[0]
